src/libopensc/card-openpgp.c

```diff
@@ -1083,20 +1090,30 @@ static int
 pgp_read_blob(sc_card_t *card, pgp_blob_t *blob)
 {
 	struct pgp_priv_data *priv = DRVDATA(card);
+	static u8 *buffer = NULL;
+	static size_t buf_size = 0;
 
 	if (blob->data != NULL)
 		return SC_SUCCESS;
 	if (blob->info == NULL)
 		return blob->status;
 
+	/* calculate buffer size required and allocate static buffer */
+	if (buffer == NULL) {
+		buf_size = MAX(priv->max_specialDO_size, 2048);
+		buffer = calloc(buf_size, sizeof(u8));
```

Comment:
When is the buffer `free`d?

Reply:
Because buffer is a `static` variable.
The buffer is not freed, because it is only allocated once during the whole lifetime of the process (see other comment).

I consider this strategy safer than the use of a dynamic buffer that needs to get allocated & freed with every call to that function, which would risk failing to allocate the memory with every call.
In this implementation, the buffer is allocated once, and every subsequent call, reuses it without any risk of failing to allocate the memory after the first call.

Reply:
I overlooked the `static` attribute. Why don't you just use plain old `malloc` and `free`? I don't see a reason for taking shortcuts here.